stan/math/fwd/mat/functor/jacobian.hpp

```diff
@@ -11,6 +11,30 @@ template <typename T, typename F>
 void jacobian(const F& f, const Eigen::Matrix<T, Eigen::Dynamic, 1>& x,
               Eigen::Matrix<T, Eigen::Dynamic, 1>& fx,
               Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& J) {
+  std::cout << "using fwd mode" << std::endl;
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  Matrix<fvar<T>, Dynamic, 1> x_fvar(x.size());
+  for (int i = 0; i < x.size(); ++i) {
+    for (int k = 0; k < x.size(); ++k)
+      x_fvar(k) = fvar<T>(x(k), i == k);
+    Matrix<fvar<T>, Dynamic, 1> fx_fvar = f(x_fvar);
+    if (i == 0) {
+      J.resize(fx_fvar.size(), x.size());
+      fx.resize(fx_fvar.size());
+      for (int k = 0; k < fx_fvar.size(); ++k)
+        fx(k) = fx_fvar(k).val_;
+    }
+    for (int k = 0; k < fx_fvar.size(); ++k) {
+      J(k, i) = fx_fvar(k).d_;
+    }
+  }
+}
+
+template <typename T, typename F>
+void jacobian_fwd(const F& f, const Eigen::Matrix<T, Eigen::Dynamic, 1>& x,
```

Comment:
Most changes to `Jacobian()` are accidental. The goal of the final PR will not be to edit these files. I put in print statements, which I later removed, when debugging the code.

There are two Jacobian functions: one uses reverse mode and the other forward mode. When we need to Jacobian of a map with a low-dimensional input and a high-dimensional output, forward mode is much more efficient. This is relevant when differentiating the covariance matrix, or the algebraic solver.

I created `jacoain_fwd()` because the code defaulted to the reverse mode version of `jacobian()`. I need to check with Bob if there's a better workaround, though this hack is pretty harmless.